### PR TITLE
Convert EpoxyRecyclerView to Kotlin

### PR DESCRIPTION
**Pull request by elihart**
In the spirit of Google's announcement of Android going Kotlin first, this converts EpoxyRecyclerView. The main benefit is that this brings built in support for the `withModels` function to allow building models with a lambda with EpoxyController receiver.

Fixes confusing cases like https://github.com/airbnb/epoxy/issues/741

The sample and wiki have been updated to reflect this new built in support.

I have also removed the beta warning from this class.

@ngsilverman @BenSchwab

### CODE REVIEW

**Comment by shaishavgandhi05** [comment]
This is great to see! Since this is done as a minor version upgrade (I'm assuming), I'd suggest first adding the [japicmp plugin](https://github.com/siom79/japicmp) to the project. It basically makes sure that all the public API of the library is consistent. This is the approach that OkHttp and Okio both used while converting to Kotlin.

**Comment by elihart** [comment]
Good suggestion about japicmp, however I couldn't get it to work after messing around with it for a while (using Okio as an example).

I feel ok merging this for now without it, but if we do more conversions in the future it would definitely be nice to have

File: epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyRecyclerView.kt [hunk 1]
```diff
@@ -0,0 +1,551 @@
...
+    fun requestModelBuild() {
+        if (epoxyController == null) {
+            throw IllegalStateException("A controller must be set before requesting a model build.")
+        }
+
+        if (epoxyController is SimpleEpoxyController) {
+            throw IllegalStateException("Models were set with #setModels, they can not be rebuilt.")
+        }
+
+        epoxyController!!.requestModelBuild()
```
[the remark this answers is not in the document]
Comment (elihart) [reply epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyRecyclerView.kt]: good idea - I made a similar change in other places but missed this one.

**Comment by prashand** [comment]
Is there a way to access the model instance from the `withModels` extension?
```
FooListItemView {
            id(item.id)
            name(item.name)
            listener(
                   //I'd like to set the model as the listener here
            )
 }
```

Model class

```

@ModelView
class FooListItemView @JvmOverloads constructor(
        context: Context,
        attrs: AttributeSet? = null,
        defStyleAttr: Int = 0
) : LinearLayout(context, attrs, defStyleAttr), View.OnClickListener { 

        private var _listener: OnClickListener? = null

        var listener: OnClickListener?
                 @CallbackProp set(value){
                  _listener = value
                  name.setOnClickListener(_listener)
        }
        get() = _listener

        override fun onClick(v: View?) {
                //OnClick Logic
        }
}
```

**Comment by elihart** [comment]
@prashand your question doesn't seem related, but the answer is no, that is not possible.

Also, you shouldn't do it anyway
```
  listener(
                   //I'd like to set the model as the listener here
            )
```

Doing that will break things for you because of the DoNotHash behavior of CallbackProp. read up about it in the wiki, but basically the listener will never get an updated version of the model.

**Comment by prashand** [comment]
@elihart thanks, I've just taken the leap into Kotlin and Epoxy having been using Java and RecyclerView. I will check the wiki more.

**Pull request merged**